### ISSUE TITLE
Nested dicts inside configuration file do not get replaced with their configuration objects

### ISSUE THREAD

**Issue by pdeubel** [neuroevolution-ai/NeuroEvolution-CTRNN_new#80]
This maybe only related to the `CnnCtrnn` brain.

Consider this example config:

```json
{
  "environment": "Qbert-v0",
  "random_seed": 0,
  "number_generations": 20,
  "optimizer": {
    "type": "MU_ES",
    "initial_gene_range": 2,
    "mu": 30,
    "lambda_": 600,
    "mutpb": 0.8
  },
  "brain": {
    "type": "CNN_CTRNN",
    "use_bias": true,
    "ctrnn_conf": {
      "number_neurons": 10,
      "neuron_activation": "relu",
      "use_bias": true,
      "delta_t": 0.5,
      "optimize_state_boundaries": "global",
      "clipping_range_max": 1.0,
      "clipping_range_min": -1.0,
      "optimize_initial_neuron_state": false
    },
    "cnn_conf": {
      "conv_size1": 3,
      "conv_feat1": 5,
      "maxp_size1": 1,
      "maxp_stride1": 1,
      "conv_size2": 3,
      "conv_feat2": 5,
      "maxp_size2": 1,
      "maxp_stride2": 1
    }
  },
  "episode_runner": {
    "number_fitness_runs": 3
  }
}
```

When running this (with default CLI arguments) I get the error `AttributeError: 'dict' object has no attribute 'conv_size1'`. This is because the `cnn_conf` dict did not get replaced with a `configurations.ConvolutionalNNCfg` object.

I guess somewhere in the recursion in `ConfigReader._replace_dicts_with_types()` something goes wrong

**Comment by pdeubel** [comment]
Okay I am stupid, I forgot the `type` parameters inside the nested dicts. If I add `"type": "CTRNN"` to `ctrnn_conf` and `"type": "CNN"` to `cnn_conf` it works as expected, the nested dicts are transformed into objects.

Its unfortunate that no error is thrown. Would it be possible to always throw an error if the "type" parameter is missing @bjuergens ?

**Comment by bjuergens** [comment]
> Would it be possible to always throw an error if the "type" parameter is missing @bjuergens ?

hm... yes. but if we wanted to intentionally pass a dict, than how would we need to handle that? 

Maybe a cleaner solution would check if all expected sub-objects are present? Like in your example when parsing the config for "brain", there should be an exception, that says, that no expected sub-brains have been found. 

Then again: When in #77 we will change everything about config-reading anyways. So anything we do here would be meaningless anyway in a couple of weeks (months? years?)

**Comment by pdeubel** [comment]
> Then again: When in #77 we will change everything about config-reading anyways. So anything we do here would be meaningless anyway in a couple of weeks (months? years?)

Yeah I agree, lets leave it like it is for the moment and #77 will fix it somehow
